subsys/testsuite/ztest/include/ztest_assert.h

```diff
@@ -184,16 +184,31 @@ static inline void z_zassert(int cond,
 /**
  * @brief Assert that 2 memory buffers have the same contents
  *
+ * This macro calls the final memory comparison assertion macro.
+ * Using double expansion allows providing some arguments by macros that
```

Comment:
Well, I'd say this is too internal/low-level to be in the *docstring*. Mind that this stuff will appear at https://docs.zephyrproject.org . This would rather be a normal code comment (not starting with `/**`, just with `/* `, or putting this explanation in commit message would be good too.

Beyond that, it's of course would be nice to get (keep?) docs on the main symbol, I guess for this, you'd need to just do:

~~~
#define zassert_mem_equal(buf, exp, size, msg) \
	zassert_mem_equal__(buf, exp, size, msg)
~~~

Reply:
The magic in this macro is in the usage of ```...```, that means that technically you may call the first macro using any number of argument.

Your proposal would only unwarp all arguments, but it would not be possible to change its count.

It would not work for fallowing pseudocode:

```
#define ARRAY_MAGIC(a) a, sizeof(a)

...

zassert_mem_equal(exp, ARRAY_MAGIC(my_array), NULL);
```

Reply:
> Anyway, I agree that it would be nice to hide this internal macro. The question is how? Do we have some special flag set if we are using the code for documentation compilation?

Yes, we have. Let me grep the tree for it ;-).

Reply:
It's pair of:

~~~
/** @cond INTERNAL_HIDDEN */
/** @endcond */
~~~

Just grep existing headers for usage example.

Reply:
Could we use it in C code?
The code below would work?

```
#if INTERNAL_HIDDEN
/*
 * Assert that 2 memory buffers have the same contents
 *
 * This macro calls the final memory comparison assertion macro.
 * Using double expansion allows providing some arguments by macros that
 * would expand to more than one values (ANSI-C99 defines that all the macro
 * arguments have to be expanded before macro call).
 *
 * Params:
 *  ... Arguments, see @ref zassert_mem_equal__
 *            for real arguments accepted.
 */
#define zassert_mem_equal(...) \
	zassert_mem_equal__(__VA_ARGS__)

#define zassert_mem_equal__(buf, exp, size, msg, ...)                        \
	zassert_equal(memcmp(buf, exp, size), 0, #buf " not equal to " #exp, \
	msg, ##__VA_ARGS__)
#else
/**
 * @brief Assert that 2 memory buffers have the same contents
 *
 * @param buf Buffer to compare
 * @param exp Buffer with expected contents
 * @param size Size of buffers
 * @param msg Optional message to print if the assertion fails
 */
#define zassert_mem_equal(buf, exp, size, msg, ...)                        \
	zassert_equal(memcmp(buf, exp, size), 0, #buf " not equal to " #exp, \
	msg, ##__VA_ARGS__)

#endif
```

The problem is that to generate nice code and hide the internal here we need to provide different code for DOXYGEN and different for C preprocessor. I am not sure how to archieve it using doxygen ```@cond``` statement - any ideas?

Reply:
Note also that proposed modification in comparison to the previous one allows the usage of formatted string in a message.

The only problem here is documentation. I can see that in the Zephyr repository the colegue @dbkinder works a lot around the documentation. Could you help, please?